"glyph names are referenced but are missing from the glyph set" with --color_format=glyf

I'm asusing here that if I want to build a black and white font, then the way to do that is supposed to be `--color_format=glyf`.

```console
> nanoemoji --color_format=glyf --build_dir=/tmp/build ~/src/openmoji/black/svg/*.svg
...
I0821 13:55:20.156143 140401459275584 __init__.py:198] Pre-processing glyphs
I0821 13:55:21.644322 140401459275584 __init__.py:201] Running DecomposeComponentsFilter on AnEmojiFamily-Regular
I0821 13:55:21.659806 140401459275584 __init__.py:201] Running CubicToQuadraticFilter on AnEmojiFamily-Regular
I0821 13:55:23.812396 140401459275584 cubicToQuadratic.py:48] New spline lengths: 1: 4989, 2: 2869, 3: 445, 4: 136, 5: 8, 6: 2
I0821 13:55:23.812526 140401459275584 __init__.py:217] Building OpenType tables
W0821 13:55:23.826405 140401459275584 ttGlyphPen.py:108] skipped non-existing component 'g_1f171.0'
W0821 13:55:23.826629 140401459275584 ttGlyphPen.py:108] skipped non-existing component 'g_1f171.0'
W0821 13:55:23.826847 140401459275584 ttGlyphPen.py:108] skipped non-existing component 'g_1f171.0'
W0821 13:55:23.827029 140401459275584 ttGlyphPen.py:108] skipped non-existing component 'g_1f171.0'
W0821 13:55:23.827389 140401459275584 ttGlyphPen.py:108] skipped non-existing component 'g_1f171.0'
W0821 13:55:23.828286 140401459275584 ttGlyphPen.py:108] skipped non-existing component 'g_1f171.0'
W0821 13:55:23.828396 140401459275584 ttGlyphPen.py:108] skipped non-existing component 'g_1f171.0'
W0821 13:55:23.828578 140401459275584 ttGlyphPen.py:108] skipped non-existing component 'g_1f171.0'
W0821 13:55:23.828719 140401459275584 ttGlyphPen.py:108] skipped non-existing component 'g_1f171.0'
W0821 13:55:23.829014 140401459275584 ttGlyphPen.py:108] skipped non-existing component 'g_1f171.0'
W0821 13:55:23.829285 140401459275584 ttGlyphPen.py:108] skipped non-existing component 'g_1f171.0'
W0821 13:55:23.831177 140401459275584 ttGlyphPen.py:108] skipped non-existing component 'g_1f30d.0'
W0821 13:55:23.831289 140401459275584 ttGlyphPen.py:108] skipped non-existing component 'g_1f30d.0'
W0821 13:55:23.831404 140401459275584 ttGlyphPen.py:108] skipped non-existing component 'g_1f30d.0'
W0821 13:55:23.831508 140401459275584 ttGlyphPen.py:108] skipped non-existing component 'g_1f30d.0'
W0821 13:55:23.831611 140401459275584 ttGlyphPen.py:108] skipped non-existing component 'g_1f30d.0'
W0821 13:55:23.831729 140401459275584 ttGlyphPen.py:108] skipped non-existing component 'g_1f311'
W0821 13:55:23.831841 140401459275584 ttGlyphPen.py:108] skipped non-existing component 'g_1f311'
W0821 13:55:23.831951 140401459275584 ttGlyphPen.py:108] skipped non-existing component 'g_1f311'
W0821 13:55:23.832069 140401459275584 ttGlyphPen.py:108] skipped non-existing component 'g_1f315'
W0821 13:55:23.832181 140401459275584 ttGlyphPen.py:108] skipped non-existing component 'g_1f315'
W0821 13:55:23.832290 140401459275584 ttGlyphPen.py:108] skipped non-existing component 'g_1f315'
W0821 13:55:23.837227 140401459275584 ttGlyphPen.py:108] skipped non-existing component 'g_1f315'
W0821 13:55:23.841771 140401459275584 ttGlyphPen.py:108] skipped non-existing component 'g_1f315'
W0821 13:55:23.856497 140401459275584 ttGlyphPen.py:108] skipped non-existing component 'g_1f3f3_fe0f_200d_1f7e6_200d_1f30c'
W0821 13:55:23.856644 140401459275584 ttGlyphPen.py:108] skipped non-existing component 'g_1f3f3_fe0f_200d_1f7e6_200d_1f30c'
W0821 13:55:23.935787 140401459275584 ttGlyphPen.py:108] skipped non-existing component 'g_1f493.2'
W0821 13:55:23.949187 140401459275584 ttGlyphPen.py:108] skipped non-existing component 'g_1f532.0'
W0821 13:55:23.958236 140401459275584 ttGlyphPen.py:108] skipped non-existing component 'g_1f3f4_e0075_e0073_e0074_e0078_e007f'
W0821 13:55:24.004620 140401459275584 ttGlyphPen.py:108] skipped non-existing component 'g_1f7e4'
W0821 13:55:24.070451 140401459275584 ttGlyphPen.py:108] skipped non-existing component 'g_23fc.1'
W0821 13:55:24.070948 140401459275584 ttGlyphPen.py:108] skipped non-existing component 'g_1f3f4_e0075_e0073_e0074_e0078_e007f'
W0821 13:55:24.071634 140401459275584 ttGlyphPen.py:108] skipped non-existing component 'g_1f7e4'
W0821 13:55:24.071752 140401459275584 ttGlyphPen.py:108] skipped non-existing component 'g_1f7e4'
W0821 13:55:24.091329 140401459275584 ttGlyphPen.py:108] skipped non-existing component 'g_2b1f.1'
W0821 13:55:24.091886 140401459275584 ttGlyphPen.py:108] skipped non-existing component 'g_2b2e.2'
Traceback (most recent call last):
  File "/usr/lib64/python3.8/runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/lib64/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/usr/local/lib/python3.8/site-packages/nanoemoji/write_font.py", line 486, in <module>
    app.run(main)
  File "/usr/local/lib/python3.8/site-packages/absl/app.py", line 300, in run
    _run_main(main, args)
  File "/usr/local/lib/python3.8/site-packages/absl/app.py", line 251, in _run_main
    sys.exit(main(argv))
  File "/usr/local/lib/python3.8/site-packages/nanoemoji/write_font.py", line 480, in main
    ufo, ttfont = _generate_color_font(config, inputs)
  File "/usr/local/lib/python3.8/site-packages/nanoemoji/write_font.py", line 430, in _generate_color_font
    ttfont = _make_ttfont(config, ufo, color_glyphs)
  File "/usr/local/lib/python3.8/site-packages/nanoemoji/write_font.py", line 177, in _make_ttfont
    ttfont = ufo2ft.compileTTF(ufo, overlapsBackend="pathops")
  File "/usr/local/lib/python3.8/site-packages/ufo2ft/__init__.py", line 225, in compileTTF
    compileFeatures(
  File "/usr/local/lib/python3.8/site-packages/ufo2ft/__init__.py", line 550, in compileFeatures
    otFont = featureCompiler.compile()
  File "/usr/local/lib/python3.8/site-packages/ufo2ft/featureCompiler.py", line 125, in compile
    self.setupFeatures()
  File "/usr/local/lib/python3.8/site-packages/ufo2ft/featureCompiler.py", line 220, in setupFeatures
    featureFile = parseLayoutFeatures(self.ufo)
  File "/usr/local/lib/python3.8/site-packages/ufo2ft/featureCompiler.py", line 48, in parseLayoutFeatures
    doc = parser.parse()
  File "/usr/local/lib/python3.8/site-packages/fontTools/feaLib/parser.py", line 103, in parse
    statements.append(self.parse_feature_block_())
  File "/usr/local/lib/python3.8/site-packages/fontTools/feaLib/parser.py", line 1437, in parse_feature_block_
    self.parse_block_(block, vertical, stylisticset, size_feature, cv_feature)
  File "/usr/local/lib/python3.8/site-packages/fontTools/feaLib/parser.py", line 1623, in parse_block_
    statements.append(self.parse_substitute_())
  File "/usr/local/lib/python3.8/site-packages/fontTools/feaLib/parser.py", line 798, in parse_substitute_
    ) = self.parse_glyph_pattern_(vertical=False)
  File "/usr/local/lib/python3.8/site-packages/fontTools/feaLib/parser.py", line 420, in parse_glyph_pattern_
    gc = self.parse_glyphclass_(accept_glyphname=True)
  File "/usr/local/lib/python3.8/site-packages/fontTools/feaLib/parser.py", line 322, in parse_glyphclass_
    self.check_glyph_name_in_glyph_set(glyph)
  File "/usr/local/lib/python3.8/site-packages/fontTools/feaLib/parser.py", line 1749, in check_glyph_name_in_glyph_set
    raise FeatureLibError(
fontTools.feaLib.error.FeatureLibError: <features>:17:14: The following glyph names are referenced but are missing from the glyph set: g_1f3fb
ninja: build stopped: subcommand failed.
Traceback (most recent call last):
  File "/usr/local/bin/nanoemoji", line 11, in <module>
    load_entry_point('nanoemoji==0.3.2.dev2+gf4f3609', 'console_scripts', 'nanoemoji')()
  File "/usr/local/lib/python3.8/site-packages/nanoemoji/nanoemoji.py", line 239, in main
    app.run(_run)
  File "/usr/local/lib/python3.8/site-packages/absl/app.py", line 300, in run
    _run_main(main, args)
  File "/usr/local/lib/python3.8/site-packages/absl/app.py", line 251, in _run_main
    sys.exit(main(argv))
  File "/usr/local/lib/python3.8/site-packages/nanoemoji/nanoemoji.py", line 230, in _run
    subprocess.run(ninja_cmd, check=True)
  File "/usr/lib64/python3.8/subprocess.py", line 512, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['ninja', '-C', '/tmp/build']' returned non-zero exit status 1.
```

--- a/src/nanoemoji/write_font.py
+++ b/src/nanoemoji/write_font.py
@@ -275,6 +275,7 @@ def _draw_glyph_extents(ufo: ufoLib2.Font, glyph: Glyph):
 def _glyf_ufo(ufo, color_glyphs):
     # glyphs by reuse_key
     glyphs = {}
+    reused = set()
     for color_glyph in color_glyphs:
         logging.debug(
             "%s %s %s",
@@ -282,7 +283,7 @@ def _glyf_ufo(ufo, color_glyphs):
             color_glyph.glyph_name,
             color_glyph.transform_for_font_space(),
         )
-        parent_glyph = ufo.get(color_glyph.glyph_name)
+        parent_glyph = ufo[color_glyph.glyph_name]
         for painted_layer in color_glyph.as_painted_layers():
             # if we've seen this shape before reuse it
             reuse_key = _inter_glyph_reuse_key(painted_layer)
@@ -291,10 +292,16 @@ def _glyf_ufo(ufo, color_glyphs):
                 glyphs[reuse_key] = glyph
             else:
                 glyph = glyphs[reuse_key]
+                reused.add(glyph.name)
             parent_glyph.components.append(Component(baseGlyph=glyph.name))
 
-        # No great reason to keep single-component glyphs around
-        if len(parent_glyph.components) == 1:
+    for color_glyph in color_glyphs:
+        parent_glyph = ufo[color_glyph.glyph_name]
+        # No great reason to keep single-component glyphs around (unless reused)
+        if (
+            len(parent_glyph.components) == 1
+            and parent_glyph.components[0].baseGlyph not in reused
+        ):
             component = ufo[parent_glyph.components[0].baseGlyph]
             del ufo[component.name]
             component.unicode = parent_glyph.unicode

--- a/src/nanoemoji/write_font.py
+++ b/src/nanoemoji/write_font.py
@@ -297,6 +297,7 @@ def _glyf_ufo(ufo, color_glyphs):
         if len(parent_glyph.components) == 1:
             component = ufo[parent_glyph.components[0].baseGlyph]
             del ufo[component.name]
+            component.unicode = parent_glyph.unicode
             ufo[color_glyph.glyph_name] = component
             assert component.name == color_glyph.glyph_name
 

--- a/tests/one_rect_glyf.ttx
+++ b/tests/one_rect_glyf.ttx
@@ -18,9 +18,11 @@
     <tableVersion version="0"/>
     <cmap_format_4 platformID="0" platEncID="3" language="0">
       <map code="0x20" name=".space"/><!-- SPACE -->
+      <map code="0xe000" name="e000"/><!-- ???? -->
     </cmap_format_4>
     <cmap_format_4 platformID="3" platEncID="1" language="0">
       <map code="0x20" name=".space"/><!-- SPACE -->
+      <map code="0xe000" name="e000"/><!-- ???? -->
     </cmap_format_4>
   </cmap>
 

--- a/tests/reused_rect_glyf.ttx
+++ b/tests/reused_rect_glyf.ttx
@@ -1,0 +1,68 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<ttFont sfntVersion="\x00\x01\x00\x00">
+
+  <GlyphOrder>
+    <!-- The 'id' attribute is only for humans; it is ignored when parsed. -->
+    <GlyphID id="0" name=".notdef"/>
+    <GlyphID id="1" name=".space"/>
+    <GlyphID id="2" name="e000"/>
+    <GlyphID id="3" name="e001"/>
+    <GlyphID id="4" name="e000.0"/>
+  </GlyphOrder>
+
+  <hmtx>
+    <mtx name=".notdef" width="0" lsb="0"/>
+    <mtx name=".space" width="100" lsb="0"/>
+    <mtx name="e000" width="100" lsb="20"/>
+    <mtx name="e000.0" width="100" lsb="20"/>
+    <mtx name="e001" width="100" lsb="20"/>
+  </hmtx>
+
+  <cmap>
+    <tableVersion version="0"/>
+    <cmap_format_4 platformID="0" platEncID="3" language="0">
+      <map code="0x20" name=".space"/><!-- SPACE -->
+      <map code="0xe000" name="e000"/><!-- ???? -->
+      <map code="0xe001" name="e001"/><!-- ???? -->
+    </cmap_format_4>
+    <cmap_format_4 platformID="3" platEncID="1" language="0">
+      <map code="0x20" name=".space"/><!-- SPACE -->
+      <map code="0xe000" name="e000"/><!-- ???? -->
+      <map code="0xe001" name="e001"/><!-- ???? -->
+    </cmap_format_4>
+  </cmap>
+
+  <loca>
+    <!-- The 'loca' table will be calculated by the compiler -->
+  </loca>
+
+  <glyf>
+
+    <!-- The xMin, yMin, xMax and yMax values
+         will be recalculated by the compiler. -->
+
+    <TTGlyph name=".notdef"/><!-- contains no outline data -->
+
+    <TTGlyph name=".space"/><!-- contains no outline data -->
+
+    <TTGlyph name="e000" xMin="20" yMin="60" xMax="80" yMax="80">
+      <component glyphName="e000.0" x="0" y="0" flags="0x204"/>
+    </TTGlyph>
+
+    <TTGlyph name="e000.0" xMin="20" yMin="60" xMax="80" yMax="80">
+      <contour>
+        <pt x="20" y="80" on="1"/>
+        <pt x="20" y="60" on="1"/>
+        <pt x="80" y="60" on="1"/>
+        <pt x="80" y="80" on="1"/>
+      </contour>
+      <instructions/>
+    </TTGlyph>
+
+    <TTGlyph name="e001" xMin="20" yMin="60" xMax="80" yMax="80">
+      <component glyphName="e000.0" x="0" y="0" flags="0x204"/>
+    </TTGlyph>
+
+  </glyf>
+
+</ttFont>

--- a/tests/write_font_test.py
+++ b/tests/write_font_test.py
@@ -100,6 +100,8 @@ def test_keep_glyph_names(svgs, color_format, keep_glyph_names):
             "untouchedsvg",
             ".ttf",
         ),
+        # keep single-component composites if component reused by more than one glyph
+        (("one_rect.svg", "one_rect.svg"), "reused_rect_glyf.ttx", "glyf", ".ttf"),
     ],
 )
 def test_write_font_binary(svgs, expected_ttx, color_format, output_format):

